### ISSUE TITLE
Nesting forms, complex data structures

### ISSUE THREAD

**Issue by ivome** [christianalfoni/formsy-react#120]
I was wondering if there is a possibility of nesting forms to create complex data structures or if there is an easy way to implement that?

If the Form itself would implement the same interface as the Form element, we could nest them indefinitely and create complex hierarchical data structures with validation. Or maybe creating a special Input field is the way to go...

I tried around a little bit with creating my own Input element with sub elements, but that does not work because I don't have access to the validation rules in an external component. 
For now I created a mapping function which can handle names like entity[property] and returns the value correctly, but that is not as nice, because all the nodes have to be inside the main form component. Also I cannot add validation rules and a sub form, like when I want to make a complete address required. 
When trying to create reusable form partials I get the error: 
Uncaught Error: Form Mixin requires component to be nested in a Form

I was thinking about something like that (simplified draft): 

First create reusable form:
```js
var AddressForm = React.createClass({
[...]
render: function(){
    return (
        <Formsy.Form {...this.props}>
             <input name="street"/>
             <input name="zip"/>
             <input name="city"/>
        </Formsy.Form>
    ); 
}
});
```

And then reuse that form as a field:
```js
 <Formsy.Form>
   <input name="name"/>
   <input name="otherProperty"/>
   <AddressForm name="address"/>
</Formsy.Form>
```

Basically what I am trying to get from the form as a value is a complex data structure like that: 
```js
{
    name: 'Someclient',
    otherProperty: 123,
    address: {
        street: 'xyz',
        zip: 'xyz',
        city: 'xyz'
    }
}
```

Any thoughts on what's the best way to approach that? I think it would be a really great feature, I'd be happy to help implementing.

**Comment by kristian-puccio** [comment]
Nice! that would be very handy.

The other thing that might also work is keypaths so the input name describes the data structure.
```<input name="address.street"/>```

Both approaches would be very handy in different situations.

**Comment by ppiekarczyk** [comment]
I like this idea, but #121 and this are separate issues.

**Comment by christianalfoni** [comment]
Hi guys and thanks for the discussion!

When React 0.14 is out, with the new context handling I think we could do something like this

```js
var AddressForm = React.createClass({
[...]
render: function(){
    return (
        <Formsy.Form {...this.props}>
             <input name="street"/>
             <input name="zip"/>
             <input name="city"/>
        </Formsy.Form>
    ); 
}
});

 <Formsy.Form>
   <input name="name"/>
   <input name="otherProperty"/>
   <AddressForm name="address" extended/>
</Formsy.Form>
```
So the idea here is that the extended prop would be passed into the sub Formsy.Form, causing Formsy to return a plain DIV instead of a FORM. That way you could easily include forms in forms. But we depend on the context change to make the sub elements work I believe. I will do some tests tomorrow to verify if current context implementation can make things a bit easier for us.

But would this useful?

**Comment by ntucker** [comment]
@christianalfoni taht sounds good, for now I'm just using #121 to refer to each nested structure, but it would be nice to have some sort of simple context change so I could just say 'everything below here assumes this context by default' kind of like Handlebars.js's with thing

**Comment by ntucker** [comment]
Here's my quick hack to get nesting working that transforms . paths. It's very quick so probably needs some optimizations before inclusion. If you want I can polish it and create a PR.

```
    for (let k in model) {
      if (k.indexOf(".") >= 0) {
        let parts = k.split(".")
        let cur = model
        for (let part of parts.slice(0, -1)) {
          if (cur[part] === undefined) {
            cur[part] = {}
          }
          cur = cur[part]
        }
        cur[parts[parts.length-1]] = model[k]
        delete model[k]
      }
    }
```

**Comment by LeoIannacone** [comment]
Thank @ntucker !

@christianalfoni can you please look at it? nesting forms are really useful in many many scenario...

**Comment by christianalfoni** [comment]
Cool, let me put this in on release today. Getting to work now :-)

**Comment by ntucker** [comment]
FYI: trying to figure out how to work with arrays now. My sample code doesn't work with them currently, though I'm not sure this library is built for it at all given you can't hold state externally.

**Comment by ivome** [comment]
Here is a mapping function I wrote to support object and array values. It has a dependency on lodash atm. 

Not tested much yet, but could be used as an idea: 

```javascript
    /**
     * Maps the input values into a complex object
     * Notation for value names:
     *
     * myobject[property]
     * myobject[test][] -> array value
     * myobject[nestedobject][otherproperty]
     *
     * @param {Array} inputs
     * @return {Object}
     */
    mapInputsToValue: function(inputs){
        var values = {};
        _.each(inputs, function(value, key){
            var currentIndex = 0;
            var objectPath = [];
            // Iterate through the whole part of the obejct
            var foundPosition = key.indexOf(']', currentIndex);

            // We have simple value
            if (foundPosition === -1){
                values[key] = value;
            } else {
                // Add complex object
                var openingBracket = key.indexOf('[');
                if (openingBracket <= 0){
                    throw new Error('Invalid name for input field: ' + key);
                } else {
                    objectPath.push(key.substring(0, openingBracket));
                }

                // We have object property
                while (foundPosition !== -1){
                    // Get current attribute name
                    var attributeName = _.last(key.substring(0, foundPosition).split('['));
                    objectPath.push(attributeName);

                    foundPosition = key.indexOf(']', foundPosition + 1);
                }

                // Get object property
                var obj = values;
                for (var i = 0; i < objectPath.length - 1; i++) {
                    var n = objectPath[i];
                    if (n in obj) {
                        obj = obj[n];
                    } else {
                        obj[n] = {};
                        obj = obj[n];
                    }
                }
                obj[objectPath[objectPath.length - 1]] = value;
            }
        });
        return values;
    }
```

**Comment by kristian-puccio** [comment]
There are a few node modules that do this already. Maybe they might be
useful?

https://www.npmjs.com/package/key-path

**Comment by christianalfoni** [comment]
Hi guys,

Are you referring to using other string values? Have now implemented:

```js
<input name="address.street"/> => {address: { street: 'value' } }
```

**Comment by christianalfoni** [comment]
This is the code btw:

```js
      return Object.keys(this.model).reduce(function (mappedModel, key) {
        
        var keyArray = key.split('.');
        while (keyArray.length) {
          mappedModel[keyArray.shift()] = keyArray.length ? {} : this.model[key];
        }

        return mappedModel;

      }.bind(this), {});
```

**Comment by christianalfoni** [comment]
Just closing it for now as any more complex implementation has to be next version. Want to get new version out today :-)

**Comment by ntucker** [comment]
this won't work on arrays I don't think

**Comment by christianalfoni** [comment]
But how could it be an array? `<MyInput name=""/>` it is always a string?

**Comment by ntucker** [comment]
The structure could be an array, and y ou need to refer to that in name.
i.e.,

{nav_items: [{a:b, c:d}, {a:f, c:g}]}

PS) This doesn't seem to be called with onChange event? currentValues seesm unaltered

**Comment by christianalfoni** [comment]
Sorry, but I do not understand how the name of an input could lead to that structure? As I understand this you want to do this:

```js
<Formsy.Form>
  <MyInput name="foo"/>
  <MyInput name="address.street"/>
  <MyInput name="address.postCode"/>
</Formsy.Form>
```
Which leads to this structure:

```js
{
  foo: 'value',
  address: {
    street: 'value',
    postCode: 'value'
  }
}
```
How would you structure the inputs to give a structure that would require arrays?

**Comment by ntucker** [comment]
That's exactly my point. This doesn't allow working with a very common use case. Even tcomb-form has built in list support ableit with a terrible UI and convoluted api.

**Comment by christianalfoni** [comment]
Please describe "list support". I do not understand what a "list" is in regards of naming an input? Would be great with an example of a FORM and how that translates to:

```js
{nav_items: [{a:b, c:d}, {a:f, c:g}]}
```

How does this form look like?

**Comment by ntucker** [comment]
http://gcanti.github.io/resources/tcomb-form/playground/playground.html select 'lists' in the dropdown on the left

**Comment by ntucker** [comment]
I suggest for Array structures:

```
Multiple inputs with the same name are supported - their extracted data will always be contained in an Array when they have some submittable data, with the exception of a group of radio buttons all having the same name, which will return the selected value only.
```

**Comment by chrbala** [comment]
I'd like to add my voice on the array issue.

The syntax could be something like this:
```
< MyInput name="addresses[0]" value="123 Some Street" />
< MyInput name="addresses[1]" value="456 Different Street" />
```
So right now that results in this kind of data:
```
{
    "addresses[0]": "123 Some Street",
    "addresses[1]": "456 Different Street"
}
```
It would be better if it added the data as an array like this:
```
{"addresses":["123 Some Street","456 Different Street"]}
```

And of course the primary use case for something like this would be to add a series of objects to an array with something like this:
```
< MyInput name={"addresses[" + index + "]"} />
```
It isn't the most convenient syntax to write in this case, so maybe there is a better way of writing it. But that's my idea for how this could be implemented.

**Comment by chrbala** [comment]
Upon further research, it looks like the jQuery ajax function I am using to post data to the server expects associative data and parses it into a hierarchical JSON object.

The syntax is like this:
```
< MyInput name="addresses[0][street]" value="Some Street"/>
< MyInput name="addresses[1][street]" value="Different Street"/>
```
which resolves to:
```
{
    addresses:
        ["street": "123 Some Street", 
        "street": "456 Different Street"]
}
```
Hopefully this is helpful to those trying to get hierarchical forms working with arrays - and possibly the development of this project.

**Comment by christianalfoni** [comment]
Hi guys,

sorry for late reply. I think it makes sense to conform the name syntax to `application/x-www-form-urlencoded` type of data, which is like you describe @chrbala.

Tried finding a good lib for this, but there does not seem to exist any? But we can sure build it. So does everybody feel comfortable with:

```js
< MyInput name="addresses[0][street]" value="Some Street"/>
< MyInput name="addresses[1][street]" value="Different Street"/>

{
    addresses:
        ["street": "123 Some Street", 
        "street": "456 Different Street"]
}
```

**Comment by ntucker** [comment]
Do you mean?
```
< MyInput name="addresses[0][street]" value="Some Street"/>
< MyInput name="addresses[1][street]" value="Different Street"/>

{
    addresses:
        [{"street": "123 Some Street"}, 
        {"street": "456 Different Street"}]
}
```

**Comment by christianalfoni** [comment]
Yes, sorry, of course. Copy/paste error :-)

**Comment by chrbala** [comment]
Ah, yeah, I did mean that. I edited my above post to reflect that.

**Comment by christianalfoni** [comment]
me too,heh

**Comment by ericdfields** [comment]
I'm dealing w/ the same data structure for nested objects (rails api). I built a Formsy form that handles formatting of the data quite well. The following is not pretty, nor well-tested, but it handles nested objects (at least one-level deep) just fine, and minimizes the amount of boilerplate you have to write.

My lib:

```javascript
var React = require('react')
var Formsy = require('formsy-react');

export class FormFor extends React.Component {

  constructor(props) {
    super(props)
    this.formatChild = this.formatChild.bind(this)
  }

  formatChild(child) {
    if (typeof child == 'string') {
      return child
    }

    let entityName = this.props.entity

    if (child.props.fields_for) {
      child.props.entity = this.props.entity
      return child
    }
    if (!_.isEmpty(child.props.name)) {
      child.props.field_name = child.props.name
      if (child.props.name.match(/[\d*]/g)) {
        child.props.name = [entityName, child.props.name].join('')
      } else {
        child.props.name = [entityName, '[', child.props.name, ']'].join('')
      }
    }
    if (child.props.children) {
      child.props.children = React.Children.map(child.props.children, this.formatChild)
    }
    return child
  }

  render() {
    let children = React.Children.map( this.props.children, this.formatChild )
    return (
      <Formsy.Form {...this.props}>
        { children }
      </Formsy.Form>
    )
  }

}
```

And then in your render method:

```javascript
        <FormFor 
          entity="patient" 
          onValidSubmit={this.submit.bind(this)} 
          onValid={this.enableButton.bind(this)} 
          onInvalid={this.disableButton.bind(this)}>
                  <TextInput
                    wrapperClass="col-md-3"
                    name="first_name"  
                    label="First name:" 
                    value={ this.props.patient.first_name } 
                    required />
     </FormFor>
````

A nested object component might look like so:

```javascript
import { attributesPrefixer } from 'helpers/forms_helper'
var HiddenInput = require('components/ui/forms/hidden_input_component')
var TextInput = require('components/ui/forms/text_input_component')
var CheckboxInput = require('components/ui/forms/checkbox_input_component')
var classNames = require('classnames')

function patientPhoneNumberFieldsTemplate(field_group,i) {

  function handleRemove(i,event) {
    event.preventDefault()
    Vanda.flux.getActions('CaseManagerPatient').removePhoneNumber(i)
  }

  function prefixer(index,name) {
    return attributesPrefixer('phone_numbers', index, name)
  }

  let classes = classNames({
    'row': true,
    'inset-row': true,
    'hide': field_group.mark_for_destroy
  })
  
  return (
    <div className={classes} key={i}>
      <TextInput
        wrapperClass="col-md-3"
        name={ prefixer(i,'title') }
        label="Title:"
        placeholder="Mobile, Work, etc."
        value={ field_group.title } />

      <TextInput
        wrapperClass="col-md-3"
        name={ prefixer(i,'value') } 
        label="Number:" 
        value={ field_group.value } />

      <CheckboxInput
        wrapperClass="col-md-3"
        name={ prefixer(i, 'is_primary') }
        label="Primary number:"
        value={ field_group.is_primary } />
        
      <HiddenInput
        name={ prefixer(i, 'id') }
        value={ field_group.id } />

      <button className="btn btn-danger" onClick={ handleRemove.bind(null,i) }>Remove</button>
    </div>
  )
}


export function patientPhoneNumberFields(fields) {
  let fields = fields.map( patientPhoneNumberFieldsTemplate )

  return (
    { fields }
  )
}
```

And that super-handy attributesPrefixer up there looks like this:

```javascript
  attributesPrefixer(attributes_name,index,name) {
    return '[' + attributes_name + '_attributes][' + index + '][' + name + ']'
  },
```

**THE PAYOFF**

You get the naming convention you need to submit the form:

```
<input name="patient[first_name]" type="text" class="form-control" value="Testing" data-reactid=".0.$=11:0.0.0.3.1.$=1$=010=20:0.$=1$=011=20:0.$=1$=010=20:0.1:0">
```

Some nested inputs, out of context:

```
<input name="patient[addresses_attributes][0][id]" type="hidden" value="81822" data-reactid=".0.$=11:0.0.0.3.1.$=1$=010=20:0.$=1$=015=20:0.$=1$=011=02$fields=020=02$0=2$0:$0.$=1$=011=20:0">

<input name="patient[addresses_attributes][0][title]" type="text" class="form-control" placeholder="Home, Work, etc." value="23" data-reactid=".0.$=11:0.0.0.3.1.$=1$=010=20:0.$=1$=015=20:0.$=1$=011=02$fields=020=02$0=2$0:$0.$=1$=012=20:0.$=1$=010=20:0.1:0">

<input name="patient[addresses_attributes][0][city]" type="text" class="form-control" value="mala" data-reactid=".0.$=11:0.0.0.3.1.$=1$=010=20:0.$=1$=015=20:0.$=1$=011=02$fields=020=02$0=2$0:$0.$=1$=014=20:0.$=1$=010=20:0.1:0">
```

Etc.

Again, suits my needs, is not a solution, but maybe this helps. Happy to answer questions.

**Comment by christianalfoni** [comment]
Okay, this is now implemented, using the separate project: https://github.com/christianalfoni/form-data-to-object. Using it as a hard dependency as it is unlikely you need it for something else...

**Comment by dsteinbach** [comment]
Sorry, I am not sure if this was addressed in this issue but can I now do something like this to take advantage of "sub-validation"?

```
var AddressForm = React.createClass({
[...]
addValidAddress:function(){
   const value = this.getValue();
   value.push(this.model());
   this.setValue(value);
}
render: function(){
    return (
        <Formsy.Form {...this.props} onValidSubmit={::this.addValidAddress}>
             <input name="street"/>
             <input name="zip"/>
             <input name="city"/>
        </Formsy.Form>
    ); 
}
});
 <Formsy.Form>
   <input name="name"/>
   <input name="otherProperty"/>
   <AddressForm name="address"/>
</Formsy.Form>
```

**Comment by Semigradsky** [comment]
@dsteinbach you can not nesting forms:
http://www.w3.org/TR/2011/WD-html5-20110525/forms.html#the-form-element
> **Content model:**
Flow content, but with no form element descendants.

**Comment by klis87** [comment]
@Semigradsky Nested form could be displayed as div, then we could check whether part of a form is valid or not - this functionality is present in many form libraries, including Angular and Redux Form and it is implemented without any violation of HTML5 spec.
